File: latBolt.py
#!/usr/bin/env python3
# Import python libraries for math and plotting
import numpy as np
import matplotlib.pyplot as plt
# plot sreamlines
import plotly.figure_factory as ff
import sys
import time

## Problem Setup

# physical parameters
H = 1 #length of x grid, y grid
U0 = 1 #physical characteristic velocity (lid)
Re = 300#5000 #Reynolds number

# lattice parameters
D,Q = (2,9)

dx = 1 #x step size
dy = 1 #y step size
dt = 1 #t step size
cs = 1/np.sqrt(3)   #lattice speed of sound
rhoo = 5.00         #lattice density initial
Ma = 0.1           #lattice mach number
tau = 0.75        #lattice relaxation time

tolerance = 1e-8    # solver final residual

## Constants

# lattice parameters
uo = Ma*cs #lattice characteristic velocity
ur = U0/uo #reference velocity

omega = 1/tau               #collision frequency
alpha = cs**2*(tau-0.5)     #lattice kinematic viscosity
N = Re*alpha/uo             #lattice Re matching physical Re
nx = int(2*np.floor(N/2))   #number of nodes in x-direction
ny = nx                     #number of nodes in y-direction
print(f'grid size: {nx}x{ny}')
np.seterr(over='raise') # raise exception on numerical overrun
# physical parameters
x = np.linspace(0,H,nx) #x nodes
y = np.linspace(0,H,ny) #y nodes

# D2Q9 model parameters
w  = np.array([4/9.0, 1/9.0, 1/9.0, 1/9.0, 1/9.0, 1/36.0, 1/36.0, 1/36.0, 1/36.0]) # weight in equilibrium distribution function
cx = np.array([0, 1, 0, -1, 0, 1, -1, -1, 1]) # discrete velocity x component
cy = np.array([0, 0, 1, 0, -1, 1,  1, -1, -1]) # discrete velocity y component

## Initialization

f = np.zeros([Q,ny,nx])
feq = np.zeros([Q,ny,nx])
rho = np.ones([ny,nx])*rhoo
u = np.zeros([ny,nx])
v = np.zeros([ny,nx])
ut = np.copy(u)
vt = np.copy(v)

error = 1.0
iterations = 0

# for i in range(1,nx-1):
#     u[ny-1,i] = uo
#     v[ny-1,i] = 0.0   # redundant

# top row (y=ny-1) gets initial velocity (excluding l/r sides, x[0] and x[-1])
u[-1,1:-1] = uo
## Solving Governing Equations

def chk(name,f):
    n = f.size
    matlab = []
    for c in range(n):
        matlab += [np.float32(sys.stdin.readline()[0:-2])]

    matlab = np.array(matlab).reshape(f.shape)
    try:
        assert(np.allclose(f,matlab))
        print(name,'matches!')
    except AssertionError:
        print(f'Mismatch Error in {name}:')
        print(f'python:\n{f.shape}\n{f}\n')
        print(f'matlab:\n{matlab.shape}\n\n{matlab}')
        print(np.abs(f-matlab) > 1e-5)
        exit()

# chk('u',u)
# chk('v',v)

start = time.time()

while error > tolerance: # and iterations<1000:

    # collision
    t1 = u**2 + v**2
    # chk('t1',t1)

    # multiply every u[j,i] * cx; same for v[j,i] * cy; add together,
    # then reshape t2 back to an array with shape [Q,ny,nx]

    t2 =  (np.outer(cx,u) + np.outer(cy,v)).reshape([Q,ny,nx])
    t2t1 = 1.0 + t2*(3.0 + 4.50*t2) - 1.50*t1
    feq = np.outer(w,rho).reshape(Q,ny,nx) * t2t1
    f = omega*feq + (1-omega)*f

    # chk('t2',t2)
    # chk('feq',feq)
    # chk('f',f)

    # streaming

    # for j in range(ny):
    #     # for i in range(nx-1,0,-1): # right to left
    #     #     f[1,j,i] = f[1,j,i-1]
    #
    #     for i in range(nx-1): # left to right
    #         f[3,j,i] = f[3,j,i+1]

    f[1,:,1:] = f[1,:,:-1]
    f[3,:,0:-1] = f[3,:,1:]
    # chk('f (r->l,l->r)',f)

    # for j in range(ny-1,0,-1): # top to bottom
        # for i in range(nx):
        #     f[2,j,i] = f[2,j-1,i]

        # for i in range(nx-1,0,-1):
        #     f[5,j,i] = f[5,j-1,i-1]
        #
        # for i in range(nx-1):
        #     f[6,j,i] = f[6,j-1,i+1]

    f[2, 1:] = f[2, :-1]
    f[5, 1:, 1:] = f[5, :-1, :-1 ]
    f[6, 1:, :-1] = f[6, :-1, 1:]
    #chk('f (t->b)',f)

    # for j in range(ny-1): # bottom to top
    #     for i in range(nx):
    #         f[4,j,i] = f[4,j+1,i]
    #
    #     for i in range(nx-1):
    #         f[7,j,i] = f[7,j+1,i+1]
    #
    #     for i in range(nx-1,0,-1):
    #         f[8,j,i] = f[8,j+1,i-1]

    f[4, :-1]      = f[4, 1:]
    f[7, :-1, :-1] = f[7, 1:, 1: ]
    f[8, :-1,  1:] = f[8, 1:, :-1 ]
    #chk('f (b->t)',f)

# boundary conditions
    #for j in range(ny):
    # bounce back on west boundary
    f[1,:,0] = f[3,:,0]
    f[5,:,0] = f[7,:,0]
    f[8,:,0] = f[6,:,0]

    # bounce back on east boundary
    f[3,:,nx-1] = f[1,:,nx-1]
    f[7,:,nx-1] = f[5,:,nx-1]
    f[6,:,nx-1] = f[8,:,nx-1]

    # chk('f (ew)',f)

    # bounce back on south boundary
    #for i in range(nx):
    f[2,0] = f[4,0]
    f[5,0] = f[7,0]
    f[6,0] = f[8,0]

    #chk('f (s)',f)

    # moving lid, north boundary
    #for i in range(1,nx-1):
    rhon =        f[0, ny-1, 1:-1]+f[1, ny-1, 1:-1]+f[3, ny-1, 1:-1] \
            +2.0*(f[2, ny-1, 1:-1]+f[6, ny-1, 1:-1]+f[5, ny-1, 1:-1])
    #print(f'rhon = {rhon:.4f}')
    rhom = rhon*uo/6.0
    f[4, ny-1, 1:-1] = f[2, ny-1, 1:-1]
    f[8, ny-1, 1:-1] = f[6, ny-1, 1:-1]+rhom
    f[7, ny-1, 1:-1] = f[5, ny-1, 1:-1]-rhom

    #chk('f (n)',f)

    # rho, u, v
    # for j in range(ny):
    #     for i in range(nx):
    #         ssum = np.float64(0.0)
    #         for k in range(Q):
    #             ssum = ssum+f[k,j,i]
    #         rho[j,i] = ssum
    rho = np.sum(f,axis=0)
    #chk('rho 0',rho)

    #for i in range(nx):
    rho[ny-1] =       f[0,ny-1]+f[1,ny-1]+f[3,ny-1] \
                +2.0*(f[2,ny-1]+f[6,ny-1]+f[5,ny-1])

    #chk('rho 1',rho)

    # for i in range(1,nx):
    #     for j in range(1,ny-1):
    #         usum = 0.0
    #         vsum = 0.0
    #         for k in range(Q):
    #             usum = usum+f[k,j,i]*cx[k]
    #             vsum = vsum+f[k,j,i]*cy[k]

    usum = np.tensordot( cx, f[:, 1:-1, 1:], axes=1)
    vsum = np.tensordot( cy, f[:, 1:-1, 1:], axes=1)

    # shape = [j,i]
    u[1:-1, 1:] = usum/rho[1:-1, 1:]
    v[1:-1, 1:] = vsum/rho[1:-1, 1:]

    # chk('u',u)
    # chk('v',v)

    # error monitoring
    error = np.linalg.norm(u-ut)/(nx*ny)+np.linalg.norm(v-vt)/(nx*ny)
    ut = np.copy(u)
    vt = np.copy(v)
    iterations = iterations+1
    # if iterations % 100 == 0:
    #     print(f'iterations = {iterations}')

# Results

# convert to physical parameters
u = ur*u+1e-6
v = ur*v+1e-6
# p = cs**2*rho
print(f'Elapsed Time: {time.time()-start}')
# Correct indexing
# uu = np.flipud(np.rot90(u))
# vv = np.flipud(np.rot90(v))
# pp = flipud(rot90(p))

## Plotting
ff.create_streamline(x,y,u,v,arrow_scale=.01,density=2).show()

# plt.show()
    # set(streams,'LineWidth',1,'color',[0 0 0])
    # daspect([1 1 1])
    # xlim([0 nx])
    # xlabel('x','fontweight','bold')
    # ylim([0 ny])
    # ylabel('y','fontweight','bold')
    # axis tight
    # box on

File: latBolt.cpp
#include <fmt/format.h>
#include <fmt/ostream.h>
#include <Eigen/Dense>
#include <Eigen/CXX11/Tensor>
#include <iostream>
#include <fstream>


int main(int argc, char* argv[]) {
    using namespace Eigen;
    
    // Problem setup
    constexpr int     H = 1;   // Length of X grid, Y grid
    constexpr double U0 = 1;   // Physical characteristic velocity (lid)
    constexpr double Re = 300; // Reynold's number

    // Lattice parameters
    constexpr int      dx = 1;
    constexpr int      dy = 1;
    constexpr int      dt = 1;
    constexpr double   cs = 1/std::sqrt(3.0);
    constexpr double rhoo = 5.00;
    constexpr double   Ma = 0.1;
    constexpr double  tau = 0.75;

    // Solver convergence tolerance
    constexpr double tolerance = 1e-8;

    // Constants
    constexpr double uo = Ma*cs;                // Lattice characteristic velocity
    constexpr double ur = U0/uo;                // Reference velocity
    constexpr double omega = 1.0/tau;           // Collision frequency
    constexpr double alpha = cs*cs*(tau-0.5);   // Lattice kinematic viscosity
    constexpr double N = Re*alpha/uo;           // Lattice Re matching physical Re
    constexpr int nx = 2*std::floor(N/2.0);     // Count of nodes in X direction
    constexpr int ny = nx;                      // Count of nodes in Y direction
    VectorXd x = VectorXd::LinSpaced(nx, 0, H); // X nodes
    VectorXd y = VectorXd::LinSpaced(ny, 0, H); // Y nodes

    fmt::print("{} x {}\n", nx, ny);

    // DQ29 model parameters
    VectorXd w(9);  // Weight in equilibrium distribution function
    w << 4/9.0, 1/9.0, 1/9.0, 1/9.0, 1/9.0, 1/36.0, 1/36.0, 1/36.0, 1/36.0;
    VectorXd cx(9); // Discrete velocity X component
    cx << 0, 1, 0, -1, 0, 1, -1, -1, 1;
    VectorXd cy(9); // Discrete velocity Y component
    cy << 0, 0, 1, 0, -1, 1, 1, -1, -1;

    // Initialization
    Tensor<double,3>   f(9,nx,ny);
    Tensor<double,3> feq(9,nx,ny);

    // NOTE(Jordan): I have to use loops to set the Tensors to zero
    // because there is a bug in the Eigen setZero function for tensors.
    for(int i=0; i<nx; ++i) {
        for(int j=0; j<ny; ++j) {
            for(int k=0; k<9; ++k) {
                f(k,i,j) = 0.0;
                feq(k,i,j) = 0.0;
            }
        }
    }

    MatrixXd rho(nx, ny); rho.fill(rhoo);
    MatrixXd  u(nx, ny);     u.setZero();
    MatrixXd  v(nx, ny);     v.setZero();
    MatrixXd ut(nx, ny);    ut.setZero();
    MatrixXd vt(nx, ny);    vt.setZero();

    double error = std::numeric_limits<double>::infinity();
    int iterations = 0;

    for(int i=0; i<u.rows(); ++i) {
        u(i, u.cols()-1) = uo;
        v(i, u.cols()-1) = 0.0;
    }

    // Solve the governing equations
    while( error > tolerance /*&& iterations < 1000*/ ) {
        // Collision step
        for(int i=0; i<nx; ++i) {
            for(int j=0; j<ny; ++j) {
                const double t1 = u(i,j)*u(i,j)+v(i,j)*v(i,j);
                for(int k=0; k<9; ++k) {
                    const double t2 = u(i,j)*cx[k]+v(i,j)*cy[k];
                    feq(k,i,j) = rho(i,j)*w[k]*(1.0+3.0*t2+4.5*t2*t2-1.5*t1); // Note(Jordan): Could use horner's method here.
                    f(k,i,j) = omega*feq(k,i,j) + (1.0 - omega) * f(k,i,j);
                }
            }
        }

        // Streaming step
        for(int j=0; j<ny; ++j) {
            // Right to left
            for(int i=nx-1; i>0; --i) {
                f(1,i,j) = f(1,i-1,j);
            }
            // Left to right
            for(int i=0; i<nx-1; ++i) {
                f(3,i,j) = f(3,i+1,j);
            }
        } 

        // Top to bottom
        for(int j=ny-1; j>0; --j) {
            for(int i=0; i<nx; ++i) {
                f(2,i,j) = f(2,i,j-1);
            }
            for(int i=nx-1; i>0; --i) {
                f(5,i,j) = f(5,i-1,j-1);
            }
            for(int i=0; i<nx-1; ++i) {
                f(6,i,j) = f(6,i+1,j-1);
            }
        }

        // Bottom to top
        for(int j=0; j<ny-1; ++j) {
            for(int i=0; i<nx; ++i) {
                f(4,i,j) = f(4,i,j+1);
            }
            for(int i=0; i<nx-1; ++i) {
                f(7,i,j) = f(7,i+1,j+1);
            }
            for(int i=nx-1; i>0; --i) {
                f(8,i,j) = f(8,i-1,j+1);
            }
        }

        // East/West boundary conditions
        for(int j=0; j<ny; ++j) {
            // Bounce back on west boundary
            f(1,0,j) = f(3,0,j);
            f(5,0,j) = f(7,0,j);
            f(8,0,j) = f(6,0,j);

            // Bounce back on east boundary
            f(3,nx-1,j) = f(1,nx-1,j);
            f(7,nx-1,j) = f(5,nx-1,j);
            f(6,nx-1,j) = f(8,nx-1,j);
        }

        // Bounce back on south boundary
        for(int i=0; i<nx; ++i) {
            f(2,i,0) = f(4,i,0);
            f(5,i,0) = f(7,i,0);
            f(6,i,0) = f(8,i,0);
        }

        // Moving lid, north boundary
        for(int i=1; i<nx-1; ++i) {
            double rhon = 1*(f(0,i,ny-1) + f(1,i,ny-1) + f(3,i,ny-1)) + 
                          2*(f(2,i,ny-1) + f(6,i,ny-1) + f(5,i,ny-1));
            f(4,i,ny-1) = f(2,i,ny-1);
            f(8,i,ny-1) = f(6,i,ny-1) + rhon*uo/6.0;
            f(7,i,ny-1) = f(5,i,ny-1) - rhon*uo/6.0;
        }


        // Update rho
        // NOTE(Jordan): Could maybe replace with: "rho = f.sum(0);"?
        for(int j=0; j<ny; ++j) {
            for(int i=0; i<nx; ++i) {
                rho(i,j) = 0.0;
                for(int k=0; k<9; ++k) {
                    rho(i,j) += f(k,i,j);
                }
            }
        }
        for(int i=0; i<nx; ++i) {
            rho(i,ny-1) = 1*( f(0,i,ny-1) + f(1,i,ny-1) + f(3,i,ny-1) ) +
                          2*( f(2,i,ny-1) + f(6,i,ny-1) + f(5,i,ny-1) );
        }

        // Update u, v
        for(int i=1; i<nx; ++i) {
            for(int j=1; j<ny-1; ++j) {
                double usum = 0.0;
                double vsum = 0.0;
                for(int k=0; k<9; ++k) {
                    usum += f(k,i,j)*cx[k];
                    vsum += f(k,i,j)*cy[k];
                }
                u(i,j) = usum / rho(i,j);
                v(i,j) = vsum / rho(i,j);
            }
        }

        // Error monitoring
        error = ((u-ut).norm() + (v-vt).norm()) / (nx*ny);
        ut = u;
        vt = v;
        iterations++;
        if( iterations % 100 == 0 ) {
            fmt::print("[{}]: {:0.8f} / {}\n", iterations, error, tolerance);
        }
    }

    // Convert from lattice units to physical units.
    u = ur*u;
    v = ur*v;

    // Save the results to .csv files for plotting in python.
    std::ofstream ufile("u.csv");
    std::ofstream vfile("v.csv");
    std::ofstream xfile("x.csv");
    std::ofstream yfile("y.csv");
    fmt::print(ufile, "{}\n", u);
    fmt::print(vfile, "{}\n", v);
    fmt::print(xfile, "{}\n", x);
    fmt::print(yfile, "{}\n", y);

    return 0;
}
